src/target/riscv/riscv-013.c

```diff
@@ -3070,14 +3100,19 @@ void write_memory_sba_simple(struct target *target, target_addr_t addr, uint32_t
 			dmi_write(target, DMI_SBADDRESS3, masked_addr);
 	}
 
-	dmi_write(target, DMI_SBDATA0, data);
+	/* Write SBDATA registers starting with highest address, since write to
+	 * SBDATA0 triggers write */
+	for (int i = write_size-1; i >= 0; i--) {
+		dmi_write(target,DMI_SBDATA0+i,write_data[i]);
+	}
 }
 
-uint32_t read_memory_sba_simple(struct target *target, target_addr_t addr, uint32_t sbcs)
+uint32_t* read_memory_sba_simple(struct target *target, target_addr_t addr,
+		uint32_t read_size, uint32_t sbcs)
 {
 	RISCV013_INFO(info);
 
-	uint32_t rd_val;
+	uint32_t* rd_val = malloc(read_size*sizeof(uint32_t));
```

Comment:
This kind of construct is really prone to memory leaks. Instead, can you have the caller pass in the pointer? That also allows you to return an error if you want.